lib/MusicBrainz/Script/RemoveUnreferencedRows.pm

```diff
@@ -0,0 +1,136 @@
+
+use List::AllUtils qw ( any );
+use MusicBrainz::Errors qw( capture_exceptions );
+use MusicBrainz::Server::Context;
+use MusicBrainz::Server::Log qw( log_info );
+
+with 'MooseX::Runnable';
+with 'MooseX::Getopt';
+with 'MusicBrainz::Script::Role::Context';
+
+has dry_run => (
+    isa => 'Bool',
+    is => 'ro',
+    default => 0,
+    traits => [ 'Getopt' ],
+    cmd_flag => 'dry-run'
+);
+
+has summary => (
+    isa => 'Bool',
+    is => 'ro',
+    default => 1,
+);
+
+has verbose => (
+    isa => 'Bool',
+    is => 'ro',
+    default => 0,
+);
+
+sub run {
+    my ($self) = @_;
+
+    my ($count, $removed) = (0, 0);
+
+    log_info { 'Checking all unreferenced rows' };
+
+    my @unreferenced_rows = @{
+        $self->c->sql->select_list_of_lists(<<~'SQL');
+              SELECT table_name,
+                     array_agg(row_id) AS row_ids
+                FROM unreferenced_row_log
+               WHERE inserted < now() - '7 day'::interval
+            GROUP BY table_name
+            SQL
+    };
+
+    for my $table_group (@unreferenced_rows) {
+        my ($table_name, $row_ids) = @$table_group;
+
+        log_info { sprintf "Checking rows for $table_name" };
+
+        my $quoted_table_name =
+            $self->c->sql->dbh->quote_identifier($table_name);
+
+        # We want to ensure the rows are still unreferenced
+        my $query = <<~"SQL";
+            SELECT id
+              FROM $quoted_table_name
+             WHERE ref_count = 0
+               AND id = any(?)
+            SQL
+
+        my $unused_ids = $self->c->sql->select_single_column_array(
+            $query,
+            $row_ids,
+        );
+
+        $count += scalar @$unused_ids;
+
+        for my $id (@$row_ids) {
+            if (any {/^$id$/} @$unused_ids) {
+                if ($self->dry_run) {
+                    log_info {
+                        sprintf "Need to remove id=%s from $table_name",
+                        $id,
+                    } if $self->verbose;
+                }
+                else {
+                    capture_exceptions(sub {
+                        Sql::run_in_transaction(sub {
+                            if ($table_name eq 'artist_credit') {
+                                $self->sql->do(<<~'SQL', $id);
+                                    DELETE FROM artist_credit_gid_redirect
+                                          WHERE new_id = ?
+                                    SQL
+                            }
+                            $self->c->sql->do(<<~"SQL", $id);
+                                DELETE FROM $quoted_table_name
+                                      WHERE id = ?
+                                SQL
+                            $self->c->sql->do(<<~'SQL', $table_name, $id);
+                                DELETE FROM unreferenced_row_log
+                                      WHERE table_name = ?
+                                        AND row_id = ?
+                                SQL
```

Comment:
This now won't run if the deletion above fails due to existing FKs (ref_count != 0), but it should.